Guard members of MonoType union & fix related bugs

kg opened the pull request:
There is a class of issue where we fail to properly check the type of a MonoType before accessing the members of its data union, which can lead to type confusion or null pointer dereference bugs.

This PR adds dedicated access helpers for every member of the union that do a type check, renames the union to make direct accesses obvious, and removes almost all direct uses of the union members.

The process of adding the access helpers and auditing related code revealed multiple bugs; I've either fixed those bugs in this PR or added FIXME comments describing the bug.

For accesses that were obviously correct (I could see a type check right above the accessor call) I used the `_unchecked` variant to optimize out the type check.

As a follow-up later on in the .NET 10 cycle we can remove the type check from these accessors to make them more efficient if we need to. I don't think we'll need to, though.

kg committed:
Guard members of MonoType union behind type check helpers

kg committed:
Reduce overhead of type access helpers

kg committed:
Cleanup and fix CI failure

kg committed:
Expand allowed cases

kg commented:
tracing\eventpipe\gcdump\gcdump\gcdump.cmd test failure is a real bug in mono.

kg committed:
Fix actual bug in bulk_type_log_single_type
Cleanup

lateralusX commented on src/mono/mono/metadata/metadata-internals.h:
```diff
@@ -1296,6 +1296,92 @@ m_type_is_byref (const MonoType *type)
 	return type->byref__;
 }
 
+MONO_NEVER_INLINE void
```
If we are going to have this in production than I believe we should adopt a similar pattern as we have when using MONO_CLASS_DEF_PRIVATE, either extend that or make a similar thing for MonoType, that way we would run all these additional checks when running ENABLE_CHECKED_BUILD_PRIVATE_TYPES (we should have CI legs doing that), but use inline wrappers with zero overhead on release builds.

kg replied:
I would prefer to do actual full checks in Release unless benchmarks show that it meaningfully regresses performance, at least during the preview period, in order to flush out any remaining issues of this type

lateralusX replied:
I would still add the same support for types as we have for class so we could do a MONO_TYPE_DEF_PRIVATE under ENABLED_CHECKED_BUILD_PRIVATE_TYPES to catch errors bypassing the getters on CI. We could also add support for the full check vs zero overhead wrappers as part of that and then just decide to initially opt into full checks during preview (and monitor perf hit during that time) and then have the option to move back to zero wrappers closer to release, but keep the full checks on CI builds running with the ENABLED_CHECKED_BUILD_PRIVATE_TYPES.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono-runtime-provider.c:
```diff
@@ -1554,8 +1554,15 @@ bulk_type_log_single_type (
-	case MONO_TYPE_ARRAY:
 	case MONO_TYPE_SZARRAY:
+	{
+		// FIXME: Previously was handled by below ARRAY block but that is incorrect; this implementation is speculative -kg
```
I assume you should be able to handle both cases similar, except that you can't assume mono_type is MonoArrayType it could also be MonoClass, so you would just need to handle the case where you get the type, rank and mono_type_parameter differently.

kg committed:
Steps to address PR feedback / cleanup

kg committed:
Cleanup and add comments

kg committed:
Add _unchecked to call sites that are obviously guarded correctly
Add comments above some apparent bugs

kg committed:
More unchecked and comment improvement for IDEs

kg committed:
More unchecked and comments

kg committed:
More unchecked

kg committed:
Attempt to fix szarray handler in eventpipe

kg committed:
Fix a bug and add comments

kg committed:
Add comment

kg committed:
Fix typo

kg commented:
I went through and audited all the union accesses I could find and changed the ones that were blatantly and correctly guarded to `_unchecked` which should remove a lot of overhead while also giving us a way to add debug/checked guards if we want them for some reason later on.

I found some more bugs (mostly around SZARRAY handling) and added FIXME comments if I couldn't trivially fix them. I will probably try to fix all the FIXMEs I've found so far.

kg committed:
Revert whitespace changes

kg committed:
Fix android and iOS build

kg committed:
Checkpoint

kg committed:
Hide the MonoType data union and introduce checked/unchecked setters

kg committed:
Fix genericinst fallthrough treating a MonoGenericClass ptr as a MonoClass ptr

kg commented:
The AOT compiler failure isn't generating usable stacks for some reason so I'll have to try and reproduce it locally in Debug configuration and hopefully get stacks there. I'm not sure if the aot compiler lacking stack traces when it crashes is intentional, I'm surprised by it.

kg committed:
Fix genericinst fallthrough treating a MonoGenericClass ptr as a MonoClass ptr, again

kg commented:
Build linux-x64 Release AllSubsets_Mono_LLVMFULLAOT_RuntimeTests llvmfullaot and its sibling look hung, but maybe they're just normally slow, or they're very slow because of the changes in this PR? I'm not familiar with these lanes and I can't tell whether the stdout on both of them stops at 1999 by coincidence or because azdo is truncating the actual output.

kg committed:
Audit more code and apply _unchecked

kg committed:
Fix type confusion in amd64 mini

kg committed:
Fix type confusion in aot-runtime-wasm

kg committed:
More audited unchecked

kg committed:
More unchecked
Prune a dead goto to make unchecked safe

kg commented:
I took a middle ground between @lateralusX 's suggestion and my personal preference, by renaming the union but not having a configuration where direct union access is allowed. Having two modes would have complicated the code a lot so I decided not to do it yet. If we decide we need direct access for some reason in a wider set of code I can do the work to enable it, but I think it's better to enforce use of the getters and setters everywhere.

Most uses of the getters ended up being unchecked because they were obviously guarded by a switch case or an if, so I expect the performance impact to be minor in practice. If this leads to unacceptable performance regressions my plan is to just disable the checks by default in release builds as we approach the later previews.

My audit revealed a handful of places where it wasn't obvious whether the union was being used safely beforehand, but also wasn't immediately obvious that it was being used *incorrectly*. As a result I couldn't "fix" them easily and left them guarded, so customers may hit failures in their applications if those locations turn out to be bugs. i.e. a function with a name indicating it's used to manipulate generic parameters that isn't doing a type check before accessing the union as if it's a generic parameter.

There's one bare use of the raw union still left in the code because the union member was being passed to some sort of macro/function as an lvalue, and I couldn't come up with a straightforward fix for that.

Something I'd love to do is have even the _unchecked accessors do checks in debug builds, just in case my audits were incorrect and I applied _unchecked somewhere I shouldn't have. But I don't know if it's worth the trouble, so I didn't bother doing it. Would love to hear people's thoughts on that.

lateralusX reviewed (commented):
Gone through all changes, mainly commenting on more places that could potentially use unchecked instead of checked gets since we already have checks of type before making calls to the types get function.

At first, I was a little reluctant to split into checked/unchecked, because it could introduce bugs by using the unchecked in wrong places or refactoring that might falsify the use of unchecked in a piece of code. Just having one implementation with full checks enabled in debug/checked build running on all our internal testing and then having a zero cost implementation in release would have mitigated this issue, but if we believe we would like to keep some of the checks in release builds, then we would need to split into two functions like you have in this PR, but then I believe it will be crucial to make the unchecked version working as the checked version in debug/checked builds so we can capture errors using the unchecked version udner false assumptions.

kg committed:
Make the _unchecked type accessors checked in debug and checked builds

Fix copy paste error

kg commented on src/mono/mono/mini/mini-generic-sharing.c:
```diff
@@ -3618,8 +3619,8 @@ is_async_method (MonoMethod *method)
 	if (attr_class && sig && ((sig->ret->type == MONO_TYPE_VOID) ||
-				(sig->ret->type == MONO_TYPE_CLASS && !strcmp (m_class_get_name (sig->ret->data.generic_class->container_class), "Task")) ||
-				(sig->ret->type == MONO_TYPE_GENERICINST && !strcmp (m_class_get_name (sig->ret->data.generic_class->container_class), "Task`1")))) {
+				(sig->ret->type == MONO_TYPE_CLASS && !strcmp (m_class_get_name (m_type_data_get_generic_class (sig->ret)->container_class), "Task")) ||
```
This one's actually a bug, if you look at it closer. Not sure what to do about it. It's calling get_generic_class on a MONO_TYPE_CLASS as a copy-paste error.

kg replied:
Think I fixed it to be what it was supposed to be. I don't think it could have possibly worked before though.

kg committed:
Address PR feedback and fix bug

kg committed:
Address PR feedback

kg commented on src/mono/mono/eventpipe/ep-rt-mono-runtime-provider.c:
```diff
@@ -1554,20 +1554,27 @@ bulk_type_log_single_type (
 	case MONO_TYPE_SZARRAY:
+	{
+		// FIXME: Previously was handled by below ARRAY block but that is incorrect; this implementation is speculative -kg
+		val->fixed_sized_data.flags |= TYPE_FLAGS_ARRAY;
```
Is the rank 1 or 0? I don't actually know how arrays work in mono. I was assuming ranks start from 0 for a single-dim array so szarray would have a rank of 0 and the `flags |= rank` would do nothing.

lateralusX replied:
I don't think the ecma spec allow rank to be 0, but the code is not explicit about it, but have not seen mono_class_create_bounded_array being called with a rank of 0. ves_icall_RuntimeType_make_array_type even force calls of rank 0 to to rank 1 when calling mono_class_create_bounded_array. Regardless I think we should just add the rank to flags for szarrays as well.

kg replied:
The original code only added the rank to flags for ARRAY and not SZARRAY. I realized while thinking this over that it might've been intentional, as a way to distinguish between SZARRAY and rank-1 ARRAY. Do you think we should include the flag anyway? I'm inclined to match the old behavior.

lateralusX replied:
ok, lets keep old behavior, this data is mainly parsed by tools like dotnet-gcdump and dotnet-pgo, so if our old way of presenting the data was working then we can keep it.

kg commented on src/mono/mono/metadata/class-init.c:
```diff
@@ -1398,8 +1399,8 @@ make_generic_param_class (MonoGenericParam *param)
 	klass->this_arg.type = t;
-	CHECKED_METADATA_WRITE_PTR ( klass->this_arg.data.generic_param ,  param );
-	CHECKED_METADATA_WRITE_PTR ( klass->_byval_arg.data.generic_param , param );
+	CHECKED_METADATA_WRITE_PTR ( klass->this_arg.data__.generic_param ,  param );
```
CHECKED_METADATA_WRITE_PTR kinda makes that impossible

kg commented on src/mono/mono/metadata/class.c:
```diff
@@ -4653,7 +4642,7 @@ mono_generic_param_get_base_type (MonoClass *klass)
 	g_assert (mono_type_is_generic_argument (type));
 
-	MonoGenericParam *gparam = type->data.generic_param;
+	MonoGenericParam *gparam = m_type_data_get_generic_param_unchecked (type);
```
mono_type_is_generic_argument ensures it, unless I misread something

lateralusX replied:
jupp, that assert should be enough.

kg commented on src/mono/mono/metadata/marshal-lightweight.c:
```diff
@@ -344,12 +344,12 @@ emit_invoke_call (MonoMethodBuilder *mb, MonoMethod *method,
 			}
 
-			t = m_class_get_byval_arg (t->data.generic_class->container_class);
+			t = m_class_get_byval_arg (m_type_data_get_generic_class (t)->container_class);
```
This switch on `type` is targeted by a goto, so there's no guarantee that `t->type == type`

kg committed:
Address PR feedback

kg committed:
Address PR feedback

kg committed:
Address PR feedback

kg commented:
Looks like one of the new set of changes broke something. Maybe the bug fix in that async bit, not sure... will have to dig in and bisect.

kg commented:
I've spent a while trying and can't reproduce any of these CI failures locally, so I'm going to have to bisect using CI...

kg committed:
Maybe fix crashes on CI

kg committed:
Fix two cases where we were partially initializing a stack-allocated MonoType instead of fully initializing it

kg commented:
I think the cause of the CI crashes and the reason I couldn't reproduce them was uninitialized stack memory. I found two places where we allocate a MonoType on the stack and don't zero the whole thing, we just fill in the fields we care about and try to use it. I accidentally removed part of that initialization in a corner case, which I think was exposing a sufficiently trashed pointer to cause us to dereference random addresses elsewhere.

kg commented on src/mono/mono/metadata/metadata.c:
```diff
@@ -2103,15 +2103,17 @@ try_get_canonical_type (MonoType *type, MonoType **canonical_type)
 
-		   klass->_byval_arg.data.klass == klass
-		   klass->this_arg.data.klass == klass
+		   klass->_byval_arg.data__.klass == klass
```
I think in this case we're talking about the underlying data structure so it's okay to not use a getter/setter in this comment.

kg commented:
I think this is ready.
llvmfullaot has one aot compile failure that I think is probably latent and nobody noticed:
```
2025-02-06T14:02:08.9864554Z   Mono Ahead of Time compiler - compiling assembly /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/Regression/Regression_2/Runtime_68568.dll
2025-02-06T14:02:08.9864965Z   AOTID 62F7688C-21FA-9BEF-3EE4-F3B99EE9DDB6
2025-02-06T14:02:08.9865436Z   Can't find custom attr constructor image: /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/Regression/Regression_2/Runtime_68568.dll mtoken: 0x0a000002 due to: Method not found: void Xunit.SkipOnMonoAttribute..ctor(string)
2025-02-06T14:02:08.9866008Z   Failed to load custom attributes from method int Runtime_68568:Main () due to Method not found: void Xunit.SkipOnMonoAttribute..ctor(string)
2025-02-06T14:02:09.0067008Z   FullAOT cannot continue if there are loader errors.
2025-02-06T14:02:09.0075096Z /__w/1/s/src/mono/msbuild/aot-compile.proj(19,9): error MSB3073: The command "/__w/1/s/artifacts/tests/coreclr/linux.x64.Release/Tests/Core_Root/corerun /__w/1/s/artifacts/tests/coreclr/linux.x64.Release/JIT/Regression/Regression_2/Runtime_68568.dll" exited with code 1.
```
I'm fairly certain I didn't cause it.

lateralusX commented on src/mono/mono/metadata/class.c:
```diff
@@ -857,21 +857,21 @@ inflate_generic_type (MonoImage *image, MonoType *type, MonoGenericContext *cont
 		nt = mono_metadata_type_dup (image, type);
 		nt->type = MONO_TYPE_GENERICINST;
-		nt->data.generic_class = gclass;
+		m_type_data_set_generic_class (nt, gclass);
```
Won't mono_metadata_lookup_generic_class return a generic class meaning this could be unchecked? If not then maybe mono_metadata_lookup_generic_class should do that and assert if not returning a generic class.

lateralusX commented on src/mono/mono/metadata/custom-attrs.c:
```diff
@@ -389,11 +391,11 @@ load_cattr_value (MonoImage *image, MonoType *t, MonoObject **out_obj, const cha
 	case MONO_TYPE_VALUETYPE:
-		if (m_class_is_enumtype (t->data.klass)) {
-			type = mono_class_enum_basetype_internal (t->data.klass)->type;
+		if (m_class_is_enumtype (m_type_data_get_klass (t))) {
```
Could we at least put m_type_data_get_klass (t) into a local only doing call once for this block?

lateralusX commented on src/mono/mono/metadata/custom-attrs.c:
```diff
@@ -719,11 +723,11 @@ load_cattr_value_noalloc (MonoImage *image, MonoType *t, const char *p, const ch
 	case MONO_TYPE_VALUETYPE:
-		if (m_class_is_enumtype (t->data.klass)) {
-			type = mono_class_enum_basetype_internal (t->data.klass)->type;
+		if (m_class_is_enumtype (m_type_data_get_klass (t))) {
```
Could we at least put m_type_data_get_klass (t) into a local only doing call once for this block?

lateralusX commented on src/mono/mono/metadata/metadata.c:
```diff
@@ -7088,14 +7087,14 @@ mono_type_to_unmanaged (MonoType *type, MonoMarshalSpec *mspec, gboolean as_fiel
-		if (t == MONO_TYPE_CLASS && (type->data.klass == mono_defaults.multicastdelegate_class ||
-					     type->data.klass == mono_defaults.delegate_class ||
-					     m_class_get_parent (type->data.klass) == mono_defaults.multicastdelegate_class)) {
+		if (t == MONO_TYPE_CLASS && (m_type_data_get_klass (type) == mono_defaults.multicastdelegate_class ||
```
What I was meaning is that we could have multiple calls to m_type_data_get_klass(type) in this block so we could put it into local and only do one call.

lateralusX commented on src/mono/mono/metadata/sre.c:
```diff
@@ -2039,11 +2039,11 @@ encode_cattr_value (MonoAssembly *assembly, char *buffer, char *p, char **retbuf
 	case MONO_TYPE_VALUETYPE:
-		if (type->data.klass->enumtype) {
-			simple_type = mono_class_enum_basetype_internal (type->data.klass)->type;
+		if (m_type_data_get_klass (type)->enumtype) {
```
We could put result of m_type_data_get_klass(type) into local and use that in these checks to reduce calls to more "expensive" m_type_data_get_klass.

lateralusX reviewed (approved):
A couple of smaller comments, otherwise it LGTM!

kg commented:
Thanks for the additional suggestions, will do another pass.

kg committed:
Address PR feedback
